store/src/java/com/zimbra/cs/service/account/GetHAB.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <https://www.gnu.org/licenses/>.
+ * ***** END LICENSE BLOCK *****
+ */
+package com.zimbra.cs.service.account;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.zimbra.common.account.Key;
+import com.zimbra.common.account.Key.DomainBy;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.soap.AccountConstants;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.account.Account;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.account.ldap.entry.LdapDistributionList;
+import com.zimbra.soap.ZimbraSoapContext;
+import com.zimbra.soap.account.message.HABGroup;
+import com.zimbra.soap.type.ZmBoolean;
+
+/**
+ * @author zimbra
+ *
+ */
+public class GetHAB extends AccountDocumentHandler {
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see
+     * com.zimbra.soap.DocumentHandler#handle(com.zimbra.common.soap.Element,
+     * java.util.Map)
+     * 
+     */
+    @Override
+    public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+        ZimbraSoapContext zsc = getZimbraSoapContext(context);
+        Provisioning prov = Provisioning.getInstance();
+        String rootGrpId = request.getAttribute(AccountConstants.A_HAB_ROOT_GROUP_ID);
+        Account acct = getAuthenticatedAccount(zsc);
+
+        ZimbraLog.account.info("Request for HAB group:%s by: ", rootGrpId, acct);
+
+        Map<String, HABGroup> groups = new HashMap<String, HABGroup>();
+        List<HABGroup> childGrpList = new ArrayList<HABGroup>();
+        LdapDistributionList group = (LdapDistributionList) prov.getGroupBasic(Key.DistributionListBy.id, rootGrpId);
+        Set<String> members = group.getAllMembersSet();
+        
+        HABGroup grp = new HABGroup();
+        grp.setId(group.getId());
+        grp.setName(group.getName());
+        grp.setRootGroup(ZmBoolean.TRUE);
+        groups.put(group.getMail(), grp);
+        grp.setAttrs(group.getAttrs());
+
+        for (String member : members) {
+            HABGroup grpChild = new HABGroup();
+            grpChild.setParentGroupId(group.getId());
+            groups.put(member, grpChild);
+            childGrpList.add(grpChild);
+        }
+        Collections.sort(childGrpList , new SortByName());
+        grp.setChildGroups(childGrpList);
+
+        List<LdapDistributionList> lists = prov
+            .getAllHabGroups(prov.get(DomainBy.name, acct.getDomainName()));
```

Comment:
instead of searching all HABGroups under domain and then filtering groups under the root group, should query groups under the root group.